lib/handler/connect.c

```diff
@@ -52,21 +52,137 @@ struct st_connect_request_t {
         size_t next;
     } server_addresses;
     h2o_timer_t timeout;
+    struct {
+        h2o_iovec_t dest_hostname;
+        const struct st_server_address_t *dest_addr;
+    } proxy_status_next_hop;
 };
 
 #define TO_BITMASK(type, len) ((type)~(((type)1 << (sizeof(type) * 8 - (len))) - 1))
 
+static h2o_iovec_t encode_sf_string(h2o_mem_pool_t *pool, const char *s, size_t slen)
+{
+    if (slen == SIZE_MAX)
+        slen = strlen(s);
+
+    /* https://tools.ietf.org/html/rfc8941#section-3.3.3 */
+    size_t to_escape = 0;
+    for (size_t i = 0; i < slen; ++i) {
+        if (s[i] == '\\' || s[i] == '"')
+            ++to_escape;
+    }
+    char *buf = h2o_mem_alloc_pool(pool, char, slen + to_escape + 3);
+    char *d = buf;
+    *(d++) = '"';
+    for (size_t i = 0; i < slen; ++i) {
+        if (s[i] == '\\' || s[i] == '"')
+            *(d++) = '\\';
+        *(d++) = s[i];
+    }
+    *(d++) = '"';
+    *d = '\0';
+    return h2o_iovec_init(buf, slen + to_escape + 2);
+}
+
+static void _make_proxy_status_error(struct st_connect_request_t *creq,
+    const char *error_type, const char *details,
+    const char *rcode, const struct sockaddr *addr, socklen_t addrlen)
+{
+    h2o_mem_pool_t *pool = &creq->src_req->pool;
+    char addr_str[NI_MAXHOST];
+
+    if (!creq->handler->config.connect_proxy_status_enabled)
+        return;
+
+    h2o_iovec_t identity = creq->src_req->conn->ctx->globalconf->proxy_status_identity;
+    if (identity.base == NULL)
+        identity = h2o_iovec_init(H2O_STRLIT("h2o"));
+
+    h2o_iovec_t next_hop = creq->proxy_status_next_hop.dest_hostname;
+    if (addr != NULL) {
+        size_t addr_str_len = h2o_socket_getnumerichost(addr, addrlen, addr_str);
+        if (addr_str_len != SIZE_MAX) {
+            next_hop = h2o_iovec_init(addr_str, addr_str_len);
+        }
+    } else if (creq->proxy_status_next_hop.dest_addr != NULL) {
+        const struct st_server_address_t *dest_addr = creq->proxy_status_next_hop.dest_addr;
+        size_t addr_str_len = h2o_socket_getnumerichost(dest_addr->sa, dest_addr->salen, addr_str);
+        if (addr_str_len != SIZE_MAX) {
+            next_hop = h2o_iovec_init(addr_str, addr_str_len);
+        }
+    }
```

Comment:
Am I correct to assume that we are trying to send the IP address rather than the DNS name of the host? Assuming that is the case, I do not think we need to consult the name provided as string. The name is always passed to `h2o_hostinfo_getaddr` is converted to `struct sockaddr`.

Reply:
I do not know if I understand this suggestion correctly.

Yes, my intention was to populate "next-hop" with the IP when the IP address is known.  In the case when the DNS resolution fails or times out, we do not have an IP address, and it seemed to me that it would make sense to report the problematic hostname in that case in the parameter "next-hop".  The callback does not give us the hostname that could not be resolved.

That said, and I do not know if that is what you are pointing out, the hostname is something that the client supplied with their request, so we do not need to be sending it back to them if there is a problem with it.  Is that what you meant?

Reply:
Ha! I had assumed that we'd never be indicating the user-supplied hostname, because as you've wondered there's no need to do that, strictly speaking.

Aside from that, if we are to retain the hostname, I think we might want to make it a direct member is because the value is not specific to proxy-status, even though the use-case might be at the moment.

All that said, I'm not entirely sure if `next-hop` is meant to be used for exposing the IP address being resolved. Assuming that we'd be doing happy eyeballs, which IP address of the two address groups should we choose to indicate to the client?

Reply:
ad64214be removes the code that tries to populate `next-hop`, which simplifies things a lot.